### DISCUSSION TITLE
Date Range on Prisma MongoDB aggregateRaw

### DISCUSSION THREAD

**Discussion by christianledgard** [prisma/prisma#12937] [Q&A] [answered]
Hi!

I'm trying to create a RAW querie with `MongoDB` and the function `aggregateRaw`. I am trying to create a `$match` for filtering my request via a date range:

```javascript
const result = await prisma.psychotherapistSession.aggregateRaw({
          pipeline: [
            {
              $match: {
                date: {
                  $gte: new Date("2022-04-01T05:00:00.000Z"),
                  $lte: new Date("2022-04-30T05:00:00.000Z"),
                },
              },
            },
          ],
        });
```

When I try to execute the request, I get an **empty array** (_the filter is not working correctly_).  Looking at the [MongoDB documentation](https://www.mongodb.com/docs/manual/reference/operator/aggregation/month/#behavior), the correct way is to use the `ISODate()` function. Obviously, if I insert a function on my code, it would recognize it as a javascript function (not a MongoDB method).

Does anyone know how to filter correctly a date with `aggregateRaw`? 

Note: I'm using `aggregateRaw` because, after the initial filtering, I have some advance `$group` and `$lookup` that are not natively supported by `prisma`.

**Comment by christianledgard** [comment, marked as answer]
I just accomplished this with the [`$dateFromString (aggregation)`](https://www.mongodb.com/docs/manual/reference/operator/aggregation/dateFromString/).

It's a little longer, but it works! 

```javascript
const result = await prisma.psychotherapistSession.aggregateRaw({
          pipeline: [
            {
              $match: {
                $expr: {
                  $and: [
                    {
                      $gte: [
                        "$date",
                        {
                          $dateFromString: {
                            dateString: "2022-04-01T05:00:00.000Z",
                          },
                        },
                      ],
                    },
                    {
                      $lte: [
                        "$date",
                        {
                          $dateFromString: {
                            dateString: "2022-04-30T05:00:00.000Z",
                          },
                        },
                      ],
                    },
                  ],
                },
              },
            },
          ],
        });
```

Credits: [Stack Overflow Post](https://stackoverflow.com/a/52707135).

**Reply by MusaliJagadeesh** [reply]
But why do you have to convert from a string in the first place? I know this works but doesn't seem right. Any suggestion from Prisma team on why this is not working when we just use like below? Same query works fine without prisma
```
$match: {
                date: {
                  $gte: new Date("2022-04-01T05:00:00.000Z"),
                  $lte: new Date("2022-04-30T05:00:00.000Z"),
                },
              },
```

**Reply by lespons** [reply]
Mongo's comparing strings here because the Date object is being converted to a string. If you need proper Date comparison, wrap it with $date
`
pipeline: [
  {
    $match: {
      date: {
        $gte: { $date: startDate },
        $lte: { $date: endDate }
      }
    }
  }
]
`

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!

**Comment by toverux** [comment]
@christianledgard's solution works, but you do not have to pass via the aggregation language (`$expr`) to do that!

Instead of this:
```js
{ $match: { date: { $gte: startDate, $lte: endDate } } }
```

Use the Extended JSON syntax:

```js
{ $match: { date: { $gte: { $date: startDate }, $lte: { $date: endDate } } } }
```

This maps to the following BSON:

```js
{ $match: { date: { $gte: DateTime("2025-01-01 0:00:00.0 +00:00:00"), $lte: DateTime("2025-02-01 0:00:00.0 +00:00:00") } } }
```
